File: packages/agent-core/src/agent/session-mode/behaviors/design.ts
import type { Agent } from '../../..';
import { DesignModeInjector } from '../../injection/design-mode';
import { BaseSessionModeBehavior } from './base';
import type { ModeEnterContext, ModeExitContext, SessionModeInjector } from './types';

export class DesignModeBehavior extends BaseSessionModeBehavior<'design'> {
  readonly kind = 'design' as const;
  readonly outputSubdirectory = 'designs';
  readonly modeModelKey = 'design';
  readonly injectorClass = DesignModeInjector as unknown as new (agent: Agent) => SessionModeInjector;
  override readonly handoffTarget = 'plan' as const;
  override readonly supportsDesignSessions = true;

  override async onEnter(ctx: ModeEnterContext): Promise<void> {
    await super.onEnter(ctx);
    ctx.agent.sessionMode.startDesignSession(ctx.id);
  }

  /**
   * Called from SessionMode.exit() which is synchronous.
   * Design session state changes must happen inline, not deferred via await.
   */
  override onExit(ctx: ModeExitContext): void {
    // Call super without await so the continuation below runs synchronously
    // eslint-disable-next-line @typescript-eslint/no-floating-promises
    super.onExit(ctx);
    ctx.agent.sessionMode.closeCurrentDesignSession(ctx.sessionModeFilePath ?? undefined);
    if (ctx.sessionModeFilePath !== null) {
      ctx.agent.sessionMode.setLastCompletedDesignFilePath(ctx.sessionModeFilePath);
    }
  }

  /**
   * Called from SessionMode.cancel() which is synchronous.
   * Design session state changes must happen inline, not deferred via await.
   */
  override onCancel(ctx: ModeExitContext): void {
    // Call super without await so the continuation below runs synchronously
    // eslint-disable-next-line @typescript-eslint/no-floating-promises
    super.onCancel(ctx);
    ctx.agent.sessionMode.closeCurrentDesignSession();
  }
}

File: rust-ody/crates/agent-rs/src/session_mode/behaviors/design.rs
use super::do_cancel;
use super::do_enter;
use super::do_exit;
use crate::records::nested::SessionModeKind;
use crate::session_mode::types::*;
use async_trait::async_trait;

pub struct DesignModeBehavior;

#[async_trait]
impl SessionModeKindBehavior for DesignModeBehavior {
    fn kind(&self) -> SessionModeKind {
        SessionModeKind::Design
    }
    fn output_subdirectory(&self) -> &str {
        "designs"
    }
    fn mode_model_key(&self) -> &str {
        "design"
    }
    fn handoff_target(&self) -> Option<&str> {
        Some("plan")
    }
    fn supports_design_sessions(&self) -> bool {
        true
    }

    async fn on_enter(
        &self,
        ctx: &ModeEnterContext,
        sm_ctx: &dyn SessionModeContext,
    ) -> anyhow::Result<()> {
        do_enter(SessionModeKind::Design, ctx, sm_ctx).await
    }

    async fn on_exit(
        &self,
        ctx: &ModeExitContext,
        sm_ctx: &dyn SessionModeContext,
    ) -> anyhow::Result<()> {
        do_exit(SessionModeKind::Design, ctx, sm_ctx, None).await
    }

    async fn on_cancel(
        &self,
        ctx: &ModeExitContext,
        sm_ctx: &dyn SessionModeContext,
    ) -> anyhow::Result<()> {
        do_cancel(SessionModeKind::Design, ctx, sm_ctx, None).await
    }
}
